src/Symfony/Component/Translation/MessageSelector.php

```diff
@@ -51,7 +51,7 @@ class MessageSelector
      */
     public function choose($message, $number, $locale)
     {
-        $parts = explode('|', $message);
+        $parts = preg_split('/(?<!\\\\)\|/', $message)
```

Comment:
This should be enough for any case: 

```
$parts = array_map(function($value){
    return str_replace('\|','|',str_replace('\\\\','\\',$value));
},preg_split( '/(?<![^\\\\]\\\\)\|/', $message));
```
It ignores escaped pipes, ignore escaped backslashes and after split it replaces escaped pipes to pipes and escaped backslashes to backslash.

there needs to be 4 backslashes to escape to one before a pipe though, but hey, it's a way.